Run in the Git root directory

kevgo opened the pull request:
@charlierudolph @allewun 

This implements #371 

This is a "best effort" implementation for the case that the user runs "git sync" in a folder that doesn't exist on main. 

The previous implementation crashes, and is unable to continue or abort, thereby at best leaving the user in a very confusing state (in a non-existing directory, without any Git functionality), and at worst  losing stashed away changes. 

With these changes, the Git Town command at least finishes properly. The user shell is still in limbo, but the user can `cd` back into his directory manually. CD'ing in the script doesn't help because it doesn't change the parent shell. 

Unfortunately I can't test this. When you are in a folder that doesn't exist on main, and Git changes to the main branch, it physically removes that folder from the hard drive. This leaves any parent process in limbo, be it your Bash/Fish shell or the surrounding Cucumber tester. Further Git or file operations are impossible. As a second-best solution, I have documented the issue in the code base to motivate and justify the respective changes.

This is still a good change imho, since it improves the reliability and robustness of Git Town in edge cases, and in general is the right thing to do. 

Steps to reproduce the issue:
* git hack test
* mkdir foo ; cd foo ; touch bar
* git add -A ; git commit -m "test"
* git sync

allewun commented:
Would it be viable to test that this situation doesn't happen anymore? I think you can detect this situation by checking the exit code of `ls $(pwd)`

charlierudolph commented on src/helpers/helpers.sh:
```diff
@@ -36,7 +36,8 @@ source "$current_dir/environment.sh" "$@"
+export INITIAL_BRANCH_NAME=$(get_current_branch_name)
+export INITIAL_OPEN_CHANGES=$(has_open_changes)
+export HAS_REMOTE=$(has_remote_url)
+export INITIAL_DIRECTORY=$(pwd)
```
Please re-indent these or explain why you un-indented them

charlierudolph replied:
Also please put in alpha order

kevgo replied:
great points. Fixed

charlierudolph commented on src/helpers/script_helpers.sh:
```diff
@@ -109,9 +109,15 @@ function run {
     remove_step_files
+
+    # Running inside a subfolder can cause issues if that folder doesn't exist
+    # on the main branch. Hence we cd into the main branch here.
```
Comment doesn't make sense to me. I don't believe it has anything to do with the main branch and just depends if the folder is known to git yet or not. And you aren't "cd"ing into the main branch but into the root git folder

kevgo replied:
Fixed the `cd` phrasing. The problem is indeed related to the folder not existing on the main branch. If it exists in both branches, no problem.

kevgo commented:
I figured out a way to spec this. Spec added.

kevgo commented:
Spec for `git sync --all` added as well.

kevgo commented:
all comments addressed

charlierudolph commented:
I don't see the need for the `git sync --all` test. I also think this could be a single feature test. Whats your reasoning for having three separate specs?

kevgo commented:
The reason for the `git sync --all` test is mostly that one could run `git sync --all` in a folder that only exists on a particular feature branch, and we want to be sure that this works similarly well to how a normal `git sync` works. If you feel strongly against this, then I can take it out, since I don't feel strongly about keeping it in.  :)

The reason for 3 separate tests is to match the existing test structure. Do you mean this could all go into one file? Where would this file go? Directly under `features/git-sync`?

charlierudolph commented:
Okay. I still don't think we need to differentiate between with open changes and without open changes as thats not important to this test. I'm okay with two tests and think they should live at
```
features/git-sync/all_branches/folder_doesnt_exist_on_main.feature
features/git-sync/current_branch/feature_branch/folder_doesnt_exist_on_main.feature
```
Also I'd prefer the file name to not "doesnt" and instead of "does not" and to say "main branch" instead of just "main"
```
folder_does_not_exist_on_main_branch.feature
```

kevgo commented:
You got it. All comments addressed

kevgo commented:
Ping

charlierudolph commented on src/helpers/script_helpers.sh:
```diff
@@ -109,9 +109,15 @@ function run {
     preconditions "$@"
     steps > "$STEPS_FILE"
     run_steps "$STEPS_FILE" undoable
+    cd_start_dir
```
What about when we hit an error and then either need to abort and continue. I think we need to add the cd's to the steps list instead.

kevgo replied:
Great point. I have refactored the solution to make cd'ing into the root directory a real, undoable, logged step. This not only displays to the user that Git Town was cd'ing out of the current directory, it also properly goes back to the directory in all situations. Specs for all cases added.

kevgo commented:
All comments addressed

kevgo commented on src/helpers/file_helpers.sh:
```diff
@@ -3,15 +3,6 @@
 
 
-# Unique string that identifies the current directory
-temp_filename_suffix="$(pwd | tr '/' '_')"
```
These changes were necessary here to make the specs pass, so I couldn't extract them into a separate PR.

charlierudolph commented on src/helpers/folder_helpers.sh:
```diff
@@ -0,0 +1,35 @@
+function change_directory {
+  local directory=$1
+  pwd
+  run_git_command "cd $directory"
```
This is not a git command

kevgo replied:
Yeah, we have some technical drift here. I have opened #493 to address this separately, because it will result in a larger shotgun refactor.

I'd like to keep printing this command to the user, and verifying it in the commands table in Cucumber, because it is a relevant operation that both the user and the feature specs should be aware of.

charlierudolph replied:
I don't think there is technical drift. We already have a function (`run_command`) that can be used here. Just need to update our tests so when it captures what commands are run the branch is optional.

charlierudolph replied:
This is not addressed to me

kevgo replied:
I can't use `run_command` in this situation, because I'd like for it to display the Git branch here, so that this step can be listed and verified in the subsequent `Then it runs the Git commands` step.

This has several benefits:
* it is important on which branch the `cd` operation happens, since the folder cd'ed out of only exists on a particular branch
* when the `cd` command happens in the order of steps is important, because if it happens at the wrong time it leaves the file system in limbo
* it allows to document when exactly we `cd` back into this folder at the end

This means the `Then it runs the Git commands` step probably has to be renamed to `Then it runs the following commands in my workspace` to account for the fact that we now also run non-Git commands in the current workspace. 

Since this is a larger operation, I'd like to do it separately. 

I'm open to doing this differently, but I don't see a way to verify the `cd` command in the spec in another similarly clean way right now. If you see one, please outline it.

charlierudolph replied:
> I can't use `run_command` in this situation, because I'd like for it to display the Git branch here, so that this step can be listed and verified in the subsequent `Then it runs the Git commands` step.

I suggested changing the step that verifies which commands are run to make the branch optional. I don't think the `cd` should display the branch. If you agree, maybe we would want the branch name in the table to say `<none>` (I don't think the branch name being empty would work since we are drying up the branch names).

> it is important on which branch the cd operation happens, since the folder cd'ed out of only exists on a particular branch

This is true for the feature you wrote, but is not guaranteed to always be true.

kevgo replied:
I have updated #493 to reflect the outcome of our conversation. This will be addressed separately to keep this PR focussed.

charlierudolph commented on src/helpers/folder_helpers.sh:
```diff
@@ -0,0 +1,35 @@
+  local directory=$1
+  pwd
+  run_git_command "cd $directory"
+  pwd
```
Why do you print the working directory twice?

charlierudolph replied:
I don't think we need to print the working directory at all

kevgo replied:
Sorry, debugging leftovers. This was pretty tricky to test, because of disappearing folders in the middle of the spec, which left Ruby dumbfounded.

charlierudolph commented on src/helpers/helpers.sh:
```diff
@@ -36,7 +37,12 @@ source "$current_dir/environment.sh" "$@"
 source "$current_dir/configuration.sh" "$@"
 
 if [ "$(is_git_repository)" == true ]; then
+  temp_filename_suffix="$(git rev-parse --show-toplevel | tr '/' '_')"
```
```sh
temp_filename_suffix=$(git_root | tr '/' '_')
```

kevgo replied:
Good catch. Fixed

charlierudolph commented on src/helpers/folder_helpers.sh:
```diff
@@ -0,0 +1,35 @@
+
+# Returns whether the current working directory
+# is the root directory of the current Git workspace.
+function is_in_git_sub_directory {
```
Based on the comment this should be named is in `is_in_git_root_directory`. To make use of that may need to implement `echo_if_false`. I think i'd prefer that.

charlierudolph replied:
Nevermind fine as is but update the comment to something like
```
# Returns whether the current working directory
# is in a subdirectory of the current Git workspace.

kevgo replied:
Done

charlierudolph commented on features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/conflict.feature:
```diff
@@ -0,0 +1,87 @@
+Feature: git sync: syncing inside a folder that doesn't exist on the main branch (with open changes)
+
+  As a developer syncing inside a committed folder that doesn't exist on the main branch
+  I want the command to at least finish properly
```
I'd say remove "at least". I want the command to finish properly

charlierudolph replied:
This should actually point to the user value in no_conflict.feature

kevgo replied:
Both implemented

charlierudolph commented on features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature:
```diff
@@ -0,0 +1,65 @@
+Feature: git sync: syncing inside a folder that doesn't exist on the main branch (with open changes)
+
+  As a developer syncing inside a committed folder that doesn't exist on the main branch
+  I want the command to at least finish properly
```
remove "at least"

charlierudolph commented on features/git-sync/current_branch/feature_branch/folder_does_not_exist_on_main_branch/no_conflict.feature:
```diff
@@ -0,0 +1,65 @@
+      | other_feature   | local and remote | other feature commit | file2            |
+    And I am on the "current_feature" branch
+    And I have an uncommitted file
+    When I run `git sync --all` in the "new_folder" folder
```
Why is this `git sync --all`. it should be `git sync` (based on the location of the feature), which means the `other_feature` branch can be removed and `current_feature` branch can be renamed to `feature`

kevgo replied:
True. I created separate specs for `git sync` and `git sync --all` now.

charlierudolph commented on features/step_definitions/folder_steps.rb:
```diff
@@ -22,6 +22,16 @@
+
+
+Then(/^I am in the "(.+?)" folder$/) do |folder_name|
+  expect(Dir.pwd.sub git_root_folder, '[project root]').to eql "[project root]/#{folder_name}"
```
Another way to consider:
```ruby
relativeProjectPath = Pathname.new(Dir.pwd).relative_path_from Pathname.new(git_root_folder)
expect(relativeProjectPath).to.eql folder_name
```

kevgo replied:
Your version is less hacky, so I've adopted it. Good stuff!

kevgo commented:
All comments addressed

charlierudolph commented:
I don't think the `git sync --all` test is necessary. Its not sufficiently different in my mind from the `git sync` test

charlierudolph commented on src/git-sync:
```diff
@@ -46,6 +46,7 @@ function skippable {
 
 
 function steps {
+  echo_if_true "change_directory $(git_root)" "$IN_SUB_FOLDER"
```
I think this is necessary for every command that syncs the main branch (extract, hack, ship, sync, and sync-fork). Would be nice if the cd was the first and last command (ie ran before preconditions)

kevgo replied:
Agreed. I'll add it to the other commands in separate PRs. I have added checkboxes to #371 to track when all commands are addressed.

kevgo commented:
I added the `git sync --all` spec to document that the `cd` really only happens at the beginning and end, and that all the branches are sync'ed between it. But I get your point about this being redundant. I am therefore keeping only one spec. The `git sync --all` spec is more comprehensive and expressive, so I keep that one, and removed the `git sync` one. I also moved it to a higher-level folder, since this spec applies to all variations of git-sync.

kevgo commented:
All comments addressed

kevgo commented:
Ready to ship?

charlierudolph commented on src/helpers/helpers.sh:
```diff
@@ -36,7 +37,12 @@ source "$current_dir/environment.sh" "$@"
+  export UNDO_STEPS_FILE="/tmp/${PROGRAM}_undo_${temp_filename_suffix}"
   export INITIAL_BRANCH_NAME=$(get_current_branch_name)
+  export INITIAL_DIRECTORY=$(pwd)
+  export IN_SUB_FOLDER=$(is_in_git_sub_directory)
```
This should either be before all the `INITIAL_...` or after. Based on ascii it appears it should go after
```
$ irb
2.2.0 :001 > "_" < "I"
 => false 
2.2.0 :002 > "I" < "_"
 => true 
2.2.0 :003 > 

```

charlierudolph commented:
:ship: after final comment is addressed
